cuda/struct_of_arrays.cu

```diff
@@ -0,0 +1,360 @@
+soa_update_k_shared(float *posx, float *posy, float *posz, float *velx, float *vely, float *velz, float *mass) {
+    // one kernel per particle
+    // 1024 threads per particle
+    // Todo choose good names for vars
+    // Note 32 is size of warp
+    int id = blockIdx.x;
+    int other = threadIdx.x;
+    __shared__ float x_values[1024];
+    __shared__ float y_values[1024];
+    __shared__ float z_values[1024];
+
+    float x_run = 0;
+    float y_run = 0;
+    float z_run = 0;
+    const float posix = posx[id];
+    const float posiy = posy[id];
+    const float posiz = posz[id];
+
+
+    for (std::size_t j = other; j < PROBLEMSIZE; j += 1024) {
+        soa_pp_interaction(posix, posiy, posiz, posx[j], posy[j], posz[j], mass[j], &x_run, &y_run, &z_run);
+    }
+    x_values[other] = x_run;
+    y_values[other] = y_run;
+    z_values[other] = z_run;
+
+    // reduce to one
+    SYNC_THREADS;
+    int border = 512;
+    while (border != 0) {
+        if (other < border) {
+            x_values[other] += x_values[other + border];
+            y_values[other] += y_values[other + border];
+            z_values[other] += z_values[other + border];
+        }
+        border /= 2;
+        SYNC_THREADS;
+    }
+    // write back to global mem
+    if (other == 0) {
+        velx[id] += x_values[0];
+        vely[id] += y_values[0];
+        velz[id] += z_values[0];
+    }
+}
+
+
+
+
+
+__global__ void
+soa_update_t(float* posx, float* posy, float* posz, float* velx, float* vely, float* velz, float* mass) {
+    // one kernel for 1024 particles
+    // 1 thread per particle
+    // Todo choose good names for vars
+    // Note 32 is size of warp
+    int id = LINEAR_ID;
+
+    float x_run = 0;
+    float y_run = 0;
+    float z_run = 0;
+    const float posix = posx[id];
+    const float posiy = posy[id];
+    const float posiz = posz[id];
+
+    int max_iter = 1024;
+    for (std::size_t other = 0; other < PROBLEMSIZE; ++other) {
+        if (id < PROBLEMSIZE) {
+            soa_pp_interaction(posix, posiy, posiz, posx[other], posy[other], posz[other], mass[other], &x_run, &y_run, &z_run);
+        }
+    }
+    velx[id] += x_run;
+    vely[id] += y_run;
+    velz[id] += z_run;
+}
+
+
+
+__global__ void
+soa_update_t_shared(float *posx, float *posy, float *posz, float *velx, float *vely, float *velz, float *mass) {
+    // one kernel per 1024 particles
+    // 1 thread per particle
+    // Todo choose good names for vars
+    // Note 32 is size of warp
+    int id = LINEAR_ID;
+
+    float x_run = 0;
+    float y_run = 0;
+    float z_run = 0;
+    const float posix = posx[id];
+    const float posiy = posy[id];
+    const float posiz = posz[id];
+    __shared__ float temp_x[1024];
+    __shared__ float temp_y[1024];
+    __shared__ float temp_z[1024];
+    __shared__ float temp_mass[1024];
+    int max_iter = 1024;
+    for (std::size_t j = 0; j < (PROBLEMSIZE + 1023) / 1024; ++j) {
+        if (threadIdx.x + j * 1024 < PROBLEMSIZE) {
```

Comment:
Your `PROBLEMSIZE` is a multiple of 1024 so this code can be simplified. For safety, just put a `static_assert(PROBLEMSIZE % 1024 ==0);` somewhere in the code. Or `_Static_assert` if you want to stick to C.